cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1240,11 +1250,140 @@ public long getProgramRunCount(ProgramId programId) {
     return result;
   }
 
+  /**
+   * Write the start time for the upgrade if it does not exist.
+   */
+  void writeUpgradeStartTimeIfNotExist() {
+    MDSKey mdsKey = new MDSKey(RUN_COUNT_FIRST_UPGRADE_TIME);
+    if (get(mdsKey, Long.TYPE) == null) {
+      LOG.info("Start to upgrade the run count");
+      write(mdsKey, TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()));
+    }
+  }
+
+  /**
+   * Upgrade the active run records.
+   */
+  boolean upgradeActiveRunRecords(int maxRows) {
+    boolean upgradeDone = upgradeActiveRowkey(TYPE_RUN_RECORD_STARTING, maxRows / 3);
```

Comment:
this isn't great as you'll end up only doing one third of the batch, and if maxRows is 1 or 2 it will not do anything. You can probably update upgradeActiveRowkey to return the number that it updated instead of a boolean.